contracts/staking/AbstractValueDistributor.sol

```diff
@@ -0,0 +1,229 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.4;
+
+import {PRECISION} from "../utils/Globals.sol";
+
+/**
+ * @notice The AbstractValueDistributor module
+ *
+ * Contract module for distributing value among users based on their shares.
+ *
+ * The algorithm ensures that the distribution is proportional to the shares
+ * held by each user and takes into account changes in the cumulative sum over time.
+ *
+ * This contract can be used as a base contract for implementing various distribution mechanisms,
+ * such as token staking, profit sharing, or dividend distribution.
+ *
+ * It includes hooks for performing additional logic
+ * when shares are added or removed, or when value is distributed.
+ */
+abstract contract AbstractValueDistributor {
+    struct UserDistribution {
+        uint256 shares;
+        uint256 cumulativeSum;
+        uint256 owedValue;
+    }
+
+    uint256 private _totalShares;
+    uint256 private _cumulativeSum;
+    uint256 private _updatedAt;
+
+    mapping(address => UserDistribution) private _userDistributions;
+
+    event SharesAdded(address user_, uint256 amount_);
+    event SharesRemoved(address user_, uint256 amount_);
+    event ValueDistributed(address user_, uint256 amount_);
+
+    /**
+     * @notice Returns the total number of shares.
+     * @return The total number of shares.
+     */
+    function totalShares() public view returns (uint256) {
+        return _totalShares;
+    }
+
+    /**
+     * @notice Returns the cumulative sum of value that has been distributed.
+     * @return The cumulative sum of value that has been distributed.
+     */
+    function cumulativeSum() public view returns (uint256) {
+        return _cumulativeSum;
+    }
+
+    /**
+     * @notice Returns the timestamp of the last update.
+     * @return The timestamp of the last update.
+     */
+    function updatedAt() public view returns (uint256) {
+        return _updatedAt;
+    }
+
+    /**
+     * @notice Returns the distribution details for a specific user.
+     * @param user_ The address of the user.
+     * @return The distribution details including user's shares, cumulative sum and value owed.
+     */
+    function userDistribution(address user_) public view returns (UserDistribution memory) {
+        return _userDistributions[user_];
+    }
+
+    /**
+     * @notice Gets the amount of value owed to a specific user.
+     * @param user_ The address of the user.
+     * @return The total owed value to the user.
+     */
+    function getOwedValue(address user_) public view returns (uint256) {
+        UserDistribution storage userDist = _userDistributions[user_];
+
+        return
+            (userDist.shares *
+                (_getFutureCumulativeSum(block.timestamp) - userDist.cumulativeSum)) /
+            PRECISION +
+            userDist.owedValue;
+    }
```

Comment:
Maybe we should add another function that will accept the timestamp for which we want to calculate the owed value